core/src/main/scala/filodb.core/memstore/PartKeyIndex.scala

```diff
@@ -1,12 +1,228 @@
+  final val PART_ID_DV = "__partIdDv__"
+  final val PART_ID_FIELD = "__partIdField__"
+  final val START_TIME = "__startTime__"
+  final val END_TIME = "__endTime__"
+  final val PART_KEY = "__partKey__"
+  final val LABEL_LIST = s"__labelList__"
+  final val FACET_FIELD_PREFIX = "$facet_"
+  final val LABEL_LIST_FACET = FACET_FIELD_PREFIX + LABEL_LIST
+
+  final val ignoreIndexNames = HashSet(START_TIME, PART_KEY, END_TIME, PART_ID_FIELD, PART_ID_DV)
+
+  def bytesRefToUnsafeOffset(bytesRefOffset: Int): Int = bytesRefOffset + UnsafeUtils.arayOffset
+
+  private def defaultTempDir(ref: DatasetRef, shardNum: Int): File = {
+    val baseDir = new File(System.getProperty("java.io.tmpdir"))
+    val baseName = s"partKeyIndex-$ref-$shardNum-${System.currentTimeMillis()}-"
+    val tempDir = new File(baseDir, baseName)
+    tempDir
+  }
+
+  private def createTempDir(ref: DatasetRef, shardNum: Int): File = {
+    val tempDir = defaultTempDir(ref, shardNum)
+    tempDir.mkdir()
+    tempDir
+  }
+}
+
+abstract class PartKeyIndexRaw(ref: DatasetRef,
+                               shardNum: Int,
+                               schema: PartitionSchema,
+                               diskLocation: Option[File] = None,
+                               protected val lifecycleManager: Option[IndexMetadataStore] = None)
+  extends StrictLogging {
+
+  protected val queryIndexLookupLatency = Kamon.histogram("index-partition-lookup-latency",
+      MeasurementUnit.time.nanoseconds)
+    .withTag("dataset", ref.dataset)
+    .withTag("shard", shardNum)
+
+  protected val indexDiskLocation = diskLocation.map(new File(_, ref.dataset + File.separator + shardNum))
+    .getOrElse(createTempDir(ref, shardNum)).toPath
+
+  // If index rebuild is triggered or the state is Building, simply clean up the index directory and start
+  // index rebuild
+  if (
+    lifecycleManager.forall(_.shouldTriggerRebuild(ref, shardNum))
+  ) {
+    logger.info(s"Cleaning up indexDirectory=$indexDiskLocation for  dataset=$ref, shard=$shardNum")
+    Utils.deleteRecursively(indexDiskLocation.toFile) match {
+      case Success(_) => // Notify the handler that the directory is now empty
+        logger.info(s"Cleaned directory for dataset=$ref, shard=$shardNum and index directory=$indexDiskLocation")
+        notifyLifecycleListener(IndexState.Empty, System.currentTimeMillis)
+
+      case Failure(t) => // Update index state as TriggerRebuild again and rethrow the exception
+        logger.warn(s"Exception while deleting directory for dataset=$ref, shard=$shardNum " +
+          s"and index directory=$indexDiskLocation with stack trace", t)
+        notifyLifecycleListener(IndexState.TriggerRebuild, System.currentTimeMillis)
+        throw new IllegalStateException("Unable to clean up index directory", t)
+    }
+  }
+  //else {
+  // TODO here we assume there is non-empty index which we need to validate
+  //}
+
+  protected def loadIndexData[T](ctor: () => T): T = try {
+    ctor()
```

Comment:
This is moved mostly as is from the Lucene code.  The place where the Lucene Index object was created is replaced by a ctor() callback that each implementation can provide.